### ISSUE TITLE
[D3D11] I owe you guys one large Graphics feature

### ISSUE THREAD

**Issue by ghost** [xrOxygen/xray-oxygen#262]
Due to obnoxious, weird, and often arrogant behavior, I feel I owe everyone who is here something.

Although I'm done with Stalker, I still owe you guys.

Name one major feature you want implemented in Stalker.
Global illumination? Special SSAO methods? Complex post-processing effect? Framework for Physically Based Rendering? Screen space reflections?

I'm giving myself till December 31st: if I don't have it done by then, then it means people were right about me and I'll forever be a colossal douche :p 

So let it be known whatever you've always wanted.

**Comment by SevastyanDark** [comment]
Let's start simple: some optimizations in shaders code.

**Comment by MatthewKush** [comment]
@SevastyanDark No problem. I'm on it.

**Comment by MatthewKush** [comment]
Ok, I have a task for someone that has what would be considered an "old" video-card that is still able to get at least 60fps at 1920x1080.
What I want to know, is the model of your card. With that I can calculate certain features, such as this one:
We can utilize the Second-Viewport that Shoker created for a secondary purpose: render-to-texture reflections (I believe these are called "Planar reflections". I've been playing Hitman 2, it has them, they're as good as you can get without proper raytracing. Based on what kind of a GPU can minimally run Stalker on DX11, I can figure out if this feature would downgrade performance, and by how much.

I have many optimizations on the way. I'll upload what I have (so far) tomorrow.

**Comment by ForserX** [comment]
AMD R4 + Cross AMD R5
DX11 (Average settings): 19 FPS in Sun/30 FPS in Rain

**Comment by MatthewKush** [comment]
Whoa. Not good. Yeah, we definitely need as many optimizations as possible.

**Comment by SevastyanDark** [comment]
FX 4330 OC 4.3GHz + GT 3010 OC
Ultra settings(ssao not hbao): Sun: 25fps; Rain: ~30 +-2 fps

**Comment by MatthewKush** [comment]
Crap. Really bad.
I just added some optimizations a second ago. Please test.

**Comment by RainbowZerg** [comment]
Despite the fact that there are many places in renderer that probably can (and should) be optimized, graphics part isn't the main problem at all. Oxygen's renderer isn't so much different from vanilla CoP or broken in some way. As far as i know, there is something wrong with scripting system that causing framerate drops by 2-4 times in some places compared to original.

**Comment by ForserX** [comment]
Renderer system is low speed

**Comment by MatthewKush** [comment]
I actually had a decent idea. I've always bitched and moaned about re-writing the renderer...
Yes, I'm offering that. I would hope RainbowZerg would help though, as he understands certain things better than me and honestly I enjoy working with him.
I didn't understand the C++ side before, and I didn't quite understand Xray before. I'm sure I do now though, honestly.

I'd actually gladly take on such a project (I enjoy stuff like this).

I do have some questions about the engine though, which have always prevented me from attempting it:
I don't understand what the "correct" way to link the new renderer, into xrEngine. I also don't understand why parts of the rendering code are inside of xrEngine and why not all of them are inside the renderer.

If I could get some answers, and some help, I think I can pull it off.

Layout:

Clustered Forward rendering which is ridiculously fast, much faster than Xray's "Deffered Lighting".

We keep all the assets in the game exactly as they are, nothing has to be done anywhere except the renderer.

We would gain the following benefits:
*Updated look and feel for the engine
*Lots more performance for everybody, allowing people on the low-end to enjoy the high performance at normal settings, and have more cool stuff to look at potentially in the process
*MSAA and transparency work with zero issues
*Lots more sales talk and bullet-points, you get the idea.

If I'm getting a big-head just tell me to stfu.

**Comment by ForserX** [comment]
Render code in xrEngine using for xrGame call's.

**Comment by MatthewKush** [comment]
Oh, ok, so I don't necessarily have to worry about how those are written then.
I'll rename the "pbr" repo, I'll work inside there.
Have a preference as to how I name the renderer?
I don't understand DX9 as much as I do D3D10&11 (everyone was already doing that, so I decided to learn the other route).
But, with some help, we could make it similar to what it looks like you guys want anyway: DX11 but backward compatible down to DirectX 9.3-9.1, right? That sound good?

**Comment by SevastyanDark** [comment]
@MattewKush Sorry, I mean gt 1030 oc

**Comment by MatthewKush** [comment]
@SevastyanDark No worries :)

Before going hob-wild crazy and re-writing the whole renderer like I stated (I'm cutting back on the weed, I smoke waaay too much), I have some ideas on what might be causing some slowdowns. It's some mistakes I made earlier in the code, I'm going to make sure they've been fixed, and if not I'll fix them and we'll see if performance improves.

**Comment by MatthewKush** [comment]
There was also a typo affecting all of R4 (just fixed it).

**Comment by MatthewKush** [comment]
I know there's still issues (working on it now).

**Comment by MatthewKush** [comment]
Alright, I was able to reproduce the bug you guys mentioned to me. The reason I didn't see it before is because I'm on Windows 10, and I guess the latest version(s) of DXGI solve the issue.

@ForserX What version of Windows are you currently on? I'm assuming you're not on Windows 10? If you are on Windows 10 then there goes that theory.

A rewrite may indeed be necessary.

**Comment by ForserX** [comment]
Win8.1

**Comment by Xottab-DUTY** [comment]
At the current moment, engine requires unique named shader resources and constants. Feature level 9 mode can create two different objects with the one same name and this behaviour breaks the engine. So, if you want to support it, the engine's renderer should be refactored.

You also have to consider that changes should not reduce performance. One sloppy move and the performance will drop.

(https://github.com/OpenXRay/xray-16/issues/281)

**Comment by MatthewKush** [comment]
@Xottab-DUTY I am aware, and I'll make sure I don't fuck it up. I'm staying sober when I code, I'm leaving the weed for the evenings.

**Comment by MatthewKush** [comment]
I think the best option for the new renderer is to utilize "bgfx" and I can customize it for Xray: https://github.com/bkaradzic/bgfx

From their readme:

" Cross-platform, graphics API agnostic, "Bring Your Own Engine/Framework" style rendering library.

Supported rendering backends:

Direct3D 9
Direct3D 11
Direct3D 12
Metal
OpenGL 2.1
OpenGL 3.1+
OpenGL ES 2
OpenGL ES 3.1
WebGL 1.0
WebGL 2.0
Supported platforms:

Android (14+, ARM, x86, MIPS)
asm.js/Emscripten (1.25.0)
FreeBSD
iOS (iPhone, iPad, AppleTV)
Linux
MIPS Creator CI20
OSX (10.12+)
RaspberryPi
SteamLink
Windows (XP, Vista, 7, 8, 10)
UWP (Universal Windows, Xbox One)"

Sounds good to me. Features we don't really need atm, but they're all optional anyways. It scales well, there are lots of games using it. Even has OpenGl/DX9.

I wouldn't leave it as-is, I'd modify it for max performance, along with you guys (you guys are good at this stuff too, I don't mean to make it all about me).

**Comment by Xottab-DUTY** [comment]
As far as I see, `bgfx` doesn't support geometric and tesselation shaders. Only vertex, pixel and compute ones. Do you really want to reimplement this stuff?

**Comment by MatthewKush** [comment]
Xottab, you're talking about this stuff like it's really complex. It's not. That would not take long to do. I know I haven't demonstrated my abilities to people so I understand the doubt,
look, maybe I should just keep patching the existing renderers guys? 
I'm running out of ideas on how I can help out here. I really am trying.

**Comment by Xottab-DUTY** [comment]
Ah, sorry me then. I didn't worked with bgfx, so yes, I thought it would be hardie to do))

> look, maybe I should just keep patching the existing renderers guys?

@ForserX, @RainbowZerg, @Giperionn, @Vertver and others, what do you think about this?

**Comment by RainbowZerg** [comment]
I must to say that all my graphics programming experience mainly consists of editing X-Ray Engine (including my activity in xrOxygen project), some little studying of DirectX samples, reading various graphics-related articles/tutorials and not very successful attempt at writing my own DX9 renderer (DX9 in 2017, lol), so my opinion should not be considered like something very important.

As for bgfx, it seems to me too generalized with its attempt to be cross-platform, there are many unnecessary things and it wouldn't be easy to replace X-Ray's renderer backend with this. In addition, though it's not critical, I prefer the more divided (organized) code structure than "all-in-one-6000-lines-long-file". And the main question: is X-Ray's renderer so slow and/or bad designed, that it must be completely thrown away? It's somewhat messed up, but I think it's possible and would be better to clean, optimize, partially rewrite existing code. Maybe I'm wrong.

The idea of implementing backward compatibility in DX11 renderer is good, as we don't need to support any Windows version below 7. With DX9 replaced by DX11 feature levels 9.1-9.3 we can use more modern API with its features and probably write at least partially unified shaders for all feature levels.

Don't know what to say much about clustered forward rendering... I think it worth trying, but I will probably stick to work on existing renderers for now (or whenever I will be able to). I want to help with this, but I often feel depressed so don't want to take up on a such large amount of work 
in order to leave it at half or disappear for a long time.

**Comment by MatthewKush** [comment]
@SevastyanDark Awesome, I'll take a look at that now.
@RainbowZerg Don't sweat it dude, I'll take care of this.

I have a lot going on over the next 2 days, but I'm taking this seriously.
Even if it's not done by Dec 31st I'll keep working on it until it's finished.

**Comment by MatthewKush** [comment]
Alright, as usual I rushed right through the previous reply and didn't read it throughly. I'm sorry.

I agree that this new renderer idea may not have been the wisest idea. I'm starting on repairing the existing renderer. I'll keep chipping away at it, and hopefully soon we'll see some nice performance improvements.

Unfortunately a lot of these improvements have less to do with the shaders and more to do with the renderer itself, but I can fix it as long as I'm patient with it.

Repair is definitely the smarter option.

**Comment by sinaps52ru** [comment]
Очень не хватает такого же шейдера как models/model_aref, но само светящегося, как models/selflight.
I really need a shader, similar to models / model_aref, but luminous itself, like models / selflight.

**Comment by ForserX** [comment]
Не о тех шейдеров речь

**Comment by MatthewKush** [comment]
I think ForserX is just saying he wants me to focus on the task at hand.

Although I've been busy, I'm now back on this.
I'm going line by line through the shaders and re-writing them, then going into the source code and modifying the value changes.
If I uploaded what I've done so far I fear it would immediately be added, which wouldn't be good, as it's slower due to the fact that I'm having to re-arrange a variety of features. 
With that said you'll likely see some commits 24 hours from now.

**Comment by SevastyanDark** [comment]
lol

**Comment by MatthewKush** [comment]
Gotta say I don't feel all that welcome here (never have, though that doesn't excuse my bizarre outbursts). It's nothing anyone says, it's the silence whenever I ask for a response/comments.

I'm going to go with my gut and go away. No hard feelings from my end, I do get it. I'd probably do the same based on my past behavior.

I experiment a lot and learn new things when I code which is the main reason I rarely meet any deadlines. 
I'm going to fork this and do so there, maybe under a different name if I make enough changes. I'll submit pull requests of any renderer-fixes/performance fixes (soon, I truly am working on this still), but tbh I can't rewrite parts of this code without stepping on some toes, which I've done too many times already.

This I suppose means that I am indeed a douche, but whatever. I'm trying anyway.

**Comment by ForserX** [comment]
@MatthewKush: 
New year's vanity now , and all of us not to the project. We have every person is a welcome guest. Look at the other discussions. That's where the real silence is. Don't be. If you want to work, work. Do not comply with terms? And who do we have to comply with them?! Don't overreact. You have talent, go ahead and develop. If you're not interested in XRay, go ahead. But stop feeling useless. That's all I'm asking of you. I was myself...

**Comment by MatthewKush** [comment]
@ForserX Just when I thought I had few friends left, and you left me this awesome message :)

Sure thing dude, you've got it. I'm back on it.

I get mopey a lot. I have my reasons, ignore my bullshit, I always get over it, sometimes I just need a little help in the right direction, like what you just did.
I have PTSD, I don't always think rationally, and I apologize for all the problems it's caused. Thank you for being understanding as I really do want to help.

If I thought this was bad code or something I wouldn't bother helping. You guys are doing an awesome job.

/End bullshit (for now, I'll likely break down again at some point, please try and be patient with me, I am trying).

**Comment by SevastyanDark** [comment]
@MatthewKush Just ignore every bullshit at come on, and you'll surely get it.

**Comment by MatthewKush** [comment]
The performance is still really bad I hate to say. The renderer is way too slow.

I'm reading about various ways to optiomize D3D today.

I think one major change that will be necessary is getting rid of all the different deffer_base.. files, and doing all surfaces in a single shader. I know how to do it but it will take a while.

**Comment by SevastyanDark** [comment]
May be use vulkan?
Also, you can use hlsl, just with a few minor changes

**Comment by MatthewKush** [comment]
@SevastyanDark I agree with you but unfortunately I only know D3D and I hear Vulkan is a really hard to learn.
I've tried it a few times and end up confused.
D3D is nowhere near as good but it's simpler to write.

**Comment by MatthewKush** [comment]
I just observed something interesting:
I played vanilla Stalker with r3_gbuffer_opt on (default) and off.

When off, it doesn't compress the gbuffer, it uses all three render-targets. The result is an INCREASE in FPS and overall less-stuttering. I noticed that in K.D's comments about SOC, he mentioned that a bigger gbuffer could actually be a good thing.

I need help reverting the removal of the gbuffer_opt code, I simply suck at C++, HLSL is where I was taught.

I realize the new focus is on Vulkan, which I ultimately agree with, but I'd still like to chip away at the performance problems DX11 has on the shader end of things (as honestly there are still a lot, I've barely started reforming this really).
tl;dr:
If there's any chance at all at having the following programmed, I'd be grateful:
1) Add back the 3rd Gbuffer rendertarget (rt_Normal)
2) Add a new stage before combine1 - I want to do tiled lighting there, but it HAS to be a compute-shader pass
3) In order for (2) to work, I need a new constant for the constant buffer. I simply need to be able to know what type of light is being rendered (directional,point,spot) as well as the number of those lights, so another constant that stores how many lights there are of each given type.

With these changes I believe performance can go way up.
